src/IO/GEHDF5Wrapper.cxx

```diff
@@ -84,15 +90,30 @@ GEHDF5Wrapper::open(const std::string& filename)
 
     file.openFile(filename, H5F_ACC_RDONLY);
 
+    //AB  Find out of file is RDF9 or RDF10
+    H5::DataSet str_file_version = file.openDataSet("/HeaderData/RDFConfiguration/fileVersion/majorVersion");
+    str_file_version.read(&rdf_ver, H5::PredType::STD_U32LE); //AB: I have no idea if this is the right type.
+
+    // AB check if the file is compressed, error if that is the case. 
+    if (rdf_ver==9)
+    {
+        unsigned int is_compresed;
+        H5::DataSet str_file_version = file.openDataSet("/HeaderData/ListHeader/isListCompressed");
+        str_file_version.read(&is_compresed, H5::PredType::STD_U32LE); //AB: I have no idea if this is the right type.
+        if (is_compresed)
+            error("The RDF9 file is compressed, we won't be able to read it. Please uncompress it and retry. Aborting");
```

Comment:
this check will only work for list-files. You'd first have to check if it is one. (should sit in config I guess). Without it, might fall-over on non-list files.

Note that there will be a similar check for sino files though (not for norm as far as I know)

Reply:
@KrisThielemans yeah still working on this one. Thing is, quite few "is valid" tests may happen, and I am not sure if they should be here? Should I make a new function to check if the file is valid, and put there all the "if file is TYPE, check this_and_that" ?

Reply:
would make sense. In the end, we will have to be able to call this from elsewhere ("is this a GE list file?"). GEHDF5Wrapper should therefore have some members bool is_list_file() I guess. we'd then check the "signature" (minimum check on HDF5 from GE), check the config dataset on the type, set the variables, and then for each type do a consistency check. The latter probably belongs more in the respective derived classes, but as you prefer.

Reply:
So I can make:

- `GEHDF5Wrapper::is_list_file()` 
- `GEHDF5Wrapper::is_norm_file()` 
- `GEHDF5Wrapper::is_sino_file()` 
- `GEHDF5Wrapper::is_geo_file()` Maybe all these being `check_xxxx_file()`? so its clear they error check?

Then some private attributes so there is no need to rerun

- `is_sino`
- `is_list`
- `is_norm`
- `is_geo`

I do not know what you mean with "config dataset on the type".

And with the derived classes you mean `Scanner::PETMR_Signa`? Not sure which classes inherit from the wrapper. I am happy to put this wherever, just really don't know where.

Reply:
I wouldn't make `GEHDF5Wrapper::is_list_file()` etc `static`. Then it has to open the file and read stuff anyway. `GEHDF5Wrapper` is associated to a HDF5 file, so it can tell us what it is.

I thought that `ProjDataGEHDF5` et al were derived from `GEHDF5Wrapper`, but it seems they use it as a member. So that seems to imply the validity checks still have to set in `GEHDF5Wrapper` for all data-types. It's unfortunate, but ok,

Seems to make sense that `check_is_valid_projdata` et al. would throw an error with a diagnostic (use `stir::error`), while the `is_` functions would return a bool (and probably do the check but catch the error). I wouldn't bother putting in the logic to avoid rerunning. This will only be called once or twice anyway. We can always leave that optimisation for others!

> I do not know what you mean with "config dataset on the type".

do you have `HeaderData/RDFConfiguration`? should have `isListFile`. It seems nothing else of value there for this purpose. sino should have `/SegmentData/Segment2`. norm should have whatever other segment number.

> with the derived classes you mean Scanner::PETMR_Signa? 

no, I meant whoever is using `GEHDF5Wrapper`. (ProjDataGEHDF5`, `CListModeData...`)

>  using boost is OK here? to play with filenames (see if file is this type or other type), or should I try to read stuff and figure it out from HDF5 data?

you can never rely on filenames, certainly not when they come from the console. We have to read the content.

Reply:
now I understand your comment about `static` members. `GEHDF5Wrapper::check_GE_signature` is `static`, which makes sense as it can be used by others to see if it's a file it can read it.

So what about the following then:
1. move most code from `check_GE_signature` to new `static` function `check_current_signature(H5file&)`, such that we can avoid opening the file twice
2. write `check_GE_signature` in terms of that one  (open file, `check_current_signature`, all within `try...catch`)
3. call `check_current_signature` in your new `check_file`
4. remove https://github.com/UCL/STIR/blob/18249ed7f4b10512429cff3a79ff27427c29b4f7/src/IO/GEHDF5Wrapper.cxx#L218-L225 and always use `initialise_scanner_from_HDF5`.
5. add some preliminary logic in `initialise_scanner_from_HDF5` to recognise the Signa

Reply:
Doing something slightly different but in the same line. 

- `check_GE_signature()` keeps being `static`
- `check_current_signature()` has no input, assumes file has been opened by `GEHDF5Wrapper::open()`. Returns `int`, instead of `bool`. Hopefully in the future we can add an `enum` class with scanner names, for now returns `1` if Signa. That way @pw351 (or @pwadhwa351 which one to use!) can use it to do step 5 (#583).

Reply:
I can't see how `check_GE_signature(filename)` can call `check_current_signature()` in the way you suggest. It'd have to create a `GEHDF5wrapper` object and then call that, but the constructor would already be called, so there's no point anymore. I think that this solution therefore implies something like
````c++
bool check_GE_signature(filename)
{
   try { GEHDF5Wrapper testobject(filename); return true;} catch (...); return false;
}
```
would work but it constructs a whole object, which might be somewhat expensive.. In my suggestion, I'd say
```c++
bool check_GE_signature(filename)
{
  H5file file;
  // open etc checks
   try { return GEHDF5Wrapper::check_current_signature(file);} catch (...); return false;
}
```
Effect is the same, but it can be more lightweight.

I don't see why `check_current_signature()` needs to return info on scanner. It doesn't make sense with the name either...

We have an enum with scanners, it sits in `Scanner` :wink:. In any case, as long as we make sure that `Scanner::get_scanner_from_name` does the name look-up correctly (by making sure that the name listed in the RDF file is listed in the `string_list` at https://github.com/UCL/STIR/blob/18249ed7f4b10512429cff3a79ff27427c29b4f7/src/buildblock/Scanner.cxx#L347), there's very little else to do.

Summary: I like my solution better :smile:

Reply:
@KrisThielemans ah, I am fast! :) I realized my mistake and deleted my comment few minutes ago!